compiler/src/iree/compiler/Codegen/Common/GPU/GPUCheckResourceUsage.cpp

```diff
@@ -29,17 +30,24 @@ class GPUCheckResourceUsagePass final
 };
 }  // namespace
 
-static int shapedTypeStaticSize(ShapedType shapedType) {
+static int shapedTypeStaticSize(memref::AllocOp allocOp,
+                                ShapedType shapedType) {
   int allocSize = 1;
   for (auto dimSize : shapedType.getShape()) {
     if (ShapedType::isDynamic(dimSize)) continue;
     allocSize *= dimSize;
   }
   if (auto elementType =
           llvm::dyn_cast<ShapedType>(shapedType.getElementType())) {
-    allocSize *= shapedTypeStaticSize(elementType);
+    allocSize *= shapedTypeStaticSize(allocOp, elementType);
   } else {
-    allocSize *= shapedType.getElementType().getIntOrFloatBitWidth();
+    auto eltTy = shapedType.getElementType();
+    if (eltTy.isIndex()) {
+      auto mod = allocOp->getParentOfType<ModuleOp>();
+      LowerToLLVMOptions options(mod.getContext(), DataLayout(mod));
```

Comment:
This pass is shared between LLVMGPU and SPIR-V, so we cannot just bake in LLVMGPU parameters via `LowerToLLVMOptions` here. We probably want to pass this in as something controllable, much like how we handle warp size here: https://github.com/openxla/iree/blob/main/compiler/src/iree/compiler/Codegen/Common/GPU/CommonGPUPasses.h#L111.

Reply:
Well good question. Best bet is to just keep all the index types as 64-bit. If we set the DL string for NVVM module as 32-bit for shared memory pointers , hopefully LLVMs induction variable coalescing takes care of things (I have seen this work well in the past).
Here though you are storing the indices to global memory in shared memory. So you need to use 64-bit element type?

Reply:
> Here though you are storing the indices to global memory in shared memory. So you need to use 64-bit element type?

Yes, that's what I did in the new version. I hardcoded 64, see https://github.com/openxla/iree/blob/eb12672d50c8263d7e1735318f24d26a4b9cfbc2/compiler/src/iree/compiler/Codegen/LLVMGPU/Passes.cpp#L510

Reply:
(64 would require extra capabilities. Not all Vulkan capable hardware supports float64/int64 values.)

Reply:
Done